source/extensions/clusters/dynamic_forward_proxy/cluster.cc

```diff
@@ -270,41 +266,42 @@ void Cluster::addOrUpdateHost(
       ASSERT(host_info == host_map_it->second.shared_host_info_);
       ASSERT(host_map_it->second.shared_host_info_->address() !=
              host_map_it->second.logical_host_->address());
-      ENVOY_LOG(debug, "updating dfproxy cluster host address '{}'", host);
-      host_map_it->second.logical_host_->setNewAddresses(
-          host_info->address(), host_info->addressList(), dummy_lb_endpoint_);
-      return;
-    }
 
-    ENVOY_LOG(debug, "adding new dfproxy cluster host '{}'", host);
-
-    emplaced_host = host_map_
-                        .try_emplace(host, host_info,
-                                     std::make_shared<Upstream::LogicalHost>(
-                                         info(), std::string{host}, host_info->address(),
-                                         host_info->addressList(), dummy_locality_lb_endpoint_,
-                                         dummy_lb_endpoint_, nullptr, time_source_))
-                        .first->second.logical_host_;
+      ENVOY_LOG(debug, "found existing dfproxy cluster host '{}'", host);
+      emplaced_host = host_map_it->second.logical_host_;
+      founded = true;
+    } else {
+      ENVOY_LOG(debug, "adding new dfproxy cluster host '{}'", host);
+      emplaced_host = host_map_
+                          .try_emplace(host, host_info,
+                                       std::make_shared<Upstream::LogicalHost>(
+                                           info(), std::string{host}, host_info->address(),
+                                           host_info->addressList(), dummy_locality_lb_endpoint_,
+                                           dummy_lb_endpoint_, nullptr, time_source_))
+                          .first->second.logical_host_;
+    }
   }
 
   ASSERT(emplaced_host);
-  if (hosts_added == nullptr) {
-    hosts_added = std::make_unique<Upstream::HostVector>();
-  }
-  hosts_added->emplace_back(emplaced_host);
+  Upstream::HostVector hosts;
+  hosts.emplace_back(emplaced_host);
+
+  if (founded) {
+    // removing host from priorityState first, with the old address.
+    updatePriorityState({}, hosts);
+    ENVOY_LOG(debug, "updating dfproxy cluster host address '{}'", host);
+    emplaced_host->setNewAddresses(host_info->address(), host_info->addressList(),
+                                   dummy_lb_endpoint_);
+    // will add host into priorityState again, with the new address.
+  }
+  updatePriorityState(hosts, {});
```

Comment:
do we actually need to do the add/remove here?  I don't think we really use priority for DFP, and the address update grabs a lock and updates (I think) globally so I'm not sure if the update is necessary.  strict DNS appears to simply do the priority update on host addition and skip on resolution change.   
One nice way to test e2e would be to do an e2e test where we start with a bad resolved IP via cache and a short DNS refresh timeout, and verify that the update makes it to the worker threads (crib off of test/extensions/filters/http/dynamic_forward_proxy/proxy_filter_integration_test.cc tests for cache files)

Reply:
> I don't think we really use priority for DFP

yep, I do think so too. can we remove the priority entirely? if yes, it would be much simpler.

> do we actually need to do the add/remove here

yes, I do think so, we have to keep tracking the correct address in priority, otherwise,
when host is TTL expired, the old(before change) address will leaking in priority.

> One nice way to test e2e would be to do an e2e test

okay, I will have a try later, after we figure out the proper way to fix it.

Reply:
the different between logical DNS and strict DNS is that - strict DNS could remove all of the IPs from priority, but logic DNS could not.
since there is only the latest IP in logic DNS when removing the expired host:
https://github.com/envoyproxy/envoy/blob/7a9543d2813e8041a69bba708cdc3c542e0ec8f5/source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc#L268

Reply:
I worry a little about the double-call to `updatePriorityState` -- I think it means theres's a brief period with no host in the priority set. Perhaps that's ok. 

An alternative might be to check for the existing host but always create a new one for the updated IP and then make a single call to `updatePriorityState` with the new/old host.

Reply:
yeah if it's not super complicated it'd be nice to fix updatePriortySet properly so it handles a remove/add of a single host.

Reply:
okay, already updated to a single call to `updatePriorityState`.

Reply:
@zuercher can you do first pass here?  I didn't expect the single call to  updatePriorityState to require an add/remove of host instead of updateAddress and I'm not really sure which is better

Reply:
hm, not sure what happened to @zuercher on this review, but I'm wondering if it's possible to do an update while not replacing the host, still doing the address resolution change option
/wait-any

Reply:
Oh, nope, seems hard to reach it, it needs two `logicalHost`, sine we want a single call to `updatePriorityState`